WP Super Cache: Simplify the preload loop and improve stopping procedures

donnchawp opened the pull request:
This PR simplifies the preload loop by removing the "taxonomy_preload" transient option recently added.

To do this it has to do more checks on the files generated by the preload which was done in the new wpsc_is_preload_active() function.

The taxonomy preload loop doesn't depend on the site owner preloading "all" posts now either. 

The "start email" had to be rejigged because of the missing transient option.

The cancellation code will work better now and should stop the process more reliably. Sometimes it doesn't stop properly if the job is unscheduled and files not removed.

The preload notification doesn't display the current URL by default now, it lets the JS script do that.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

## Does this pull request change what data or activity we track or use?
no

## Testing instructions:
Apply the patch. Test the preload by running it, with or without taxonomies. Cancel it and make sure it cancels properly.

donnchawp committed:
Simplify the preload loop

donnchawp committed:
changelog

dilirity commented on projects/plugins/super-cache/wp-cache.php:
```diff
@@ -1011,13 +1011,12 @@ function toggleLayer( whichLayer ) {
 					$url  = file_get_contents( $cache_path . 'preload_permalink.txt' );
-					$msg .= '<p>' . sprintf( __( '<strong>Page last cached:</strong> %s', 'wp-super-cache' ), '<span id="preload_status">' . $url . '</span>' ) . '</p>';
+					// translators: %s is a URL.
+					$msg .= '<p>' . sprintf( __( '<strong>Page last cached:</strong> %s', 'wp-super-cache' ), '<div><ul id="preload_status"></ul></div>' ) . '</p>';
```
The extra div breaks the paragraph. ul is a block item so it's unnecessary.

donnchawp replied:
Thanks, removed them!

dilirity commented on projects/plugins/super-cache/wp-cache.php:
```diff
@@ -3638,32 +3684,38 @@ function wpsc_reset_preload_settings() {
+		// phpcs:ignore -- WordPress.WP.AlternativeFunctions.file_system_read_fopen WordPress.PHP.NoSilencedErrors.Discouraged
+		$fp = @fopen( $stop_preload_flag, 'w' );
+		// phpcs:ignore -- WordPress.WP.AlternativeFunctions.file_system_operations_fclose WordPress.PHP.NoSilencedErrors.Discouraged
+		@fclose( $fp );
```
Does this work? I'm not sure why it's necessary to open a file and close it :D

On second look - it looks like this creates the file if it doesn't exist.

donnchawp replied:
It's a flag that stops the preload. It just creates an empty file. :)

dilirity commented on projects/plugins/super-cache/wp-cache.php:
```diff
@@ -3638,32 +3684,38 @@ function wpsc_reset_preload_settings() {
+	$stop_preload_flag = $cache_path . 'stop_preload.txt';
+	if ( file_exists( $stop_preload_flag ) && time() - filemtime( $stop_preload_flag ) > 2 ) {
+		wp_cache_debug( 'wpsc_cancel_preload: force cancelled preload' );
+		wpsc_reset_preload_settings(); // reset the preload if we already tried more than 2 seconds ago.
```
I feel like this comment should be right above the `if` statement. Took me a while to figure out what's going on. Also - why 2 seconds?

donnchawp replied:
The 2 second wait is in case someone double tapped the "cancel" button. I'd rather the preload process stop the preload loop than the process serving the admin page.
But if the preload has got stuck or the user really wants to cancel immediately, then cancel it. That might have unforeseen consequences though - if the preload loop has started, then it will cache ten pages, and then schedule another loop.

I think it would be useful to make the preload loop more resilient. One way might be to check for the mutex file before scheduling the next loop. If it's not there it means the site owner cancelled the preload.

donnchawp replied:
That 2 second wait is now removed. When the user cancels the preload, all the settings are reset, and the "stop preload" file are created to make sure that the preload stops. I also added a check for that file to `wpsc_is_preload_active` and use that where I check for that "stop" file.

dilirity commented on projects/plugins/super-cache/wp-cache.php:
```diff
@@ -3207,113 +3207,131 @@ function wp_cron_preload_cache() {
-					$details = explode( "\n", file_get_contents( $taxonomy_filename ) );
+				if ( ! $just_started_preloading && $wp_cache_preload_email_me ) {
+					// translators: 1: site url, 2: blank space
+					wp_mail( get_option( 'admin_email' ), sprintf( __( '[%1$s] Cache Preload Started', 'wp-super-cache' ), home_url(), '' ), ' ' );
```
Why is there a blank space needed here?

donnchawp replied:
I found that wp_mail() refused to send a blank email in the past. I'm not sure if that has changed since, however.

dilirity replied:
Today I learned - thanks! I didn't know you can't use `wp_mail` with an empty message. I checked the code and it's intended. Sending an empty string results in a `WP_Error`:
```
Message body empty
```

However, I was referring to the blank space in the translators line [here](https://github.com/Automattic/jetpack/pull/31022/files/bfd32c460a106b90d7ea24e1e9af364cd7d729af#diff-49e24876fd767aa7d12d7018a180f19e42dc62b47484a137fe4d42fdb31d10f6R3237). Only the message is translated, the blank space isn't.

donnchawp replied:
oh, thanks! Will fix that.

dilirity commented on projects/plugins/super-cache/wp-cache.php:
```diff
@@ -3207,113 +3207,131 @@ function wp_cron_preload_cache() {
+		}
+	} elseif ( $c === 0 && $wp_cache_preload_email_me ) {
+		// translators: Home URL of website, blank space
+		wp_mail( get_option( 'admin_email' ), sprintf( __( '[%1$s] Cache Preload Started', 'wp-super-cache' ), home_url(), '' ), ' ' );
```
I noticed the email notification code has been added twice. Wouldn't it make more sense to send the notification when the mutex file is created? As I understand, that file is what prevents multiple preloads from happening at the same time.

donnchawp replied:
The mutex file is deleted at the end of a loop, and then recreated on the next scheduled loop, so it's impossible to know at the top of the function if this is the first or 10th loop, especially as I got rid of the taxonomy_preload transient option.

dilirity replied:
Thanks for elaborating. On a side note, the blank space isn't part of the translation string.

donnchawp replied:
Thanks, fixed that!

dilirity reviewed (changes requested):
I tried testing the preload on trunk and this branch, but I wasn't able to see much difference compared to how it works/stops. Sorry if I'm missing something.

Because almost the whole `wp_cron_preload_cache` function was modified (judging by GIT), reviewing it was kind of hard. I think we need to do something about separating "make the monorepo happy" changes from the actual update.

I left some comments on what I thought could be improved. Nothing major, just a few small things.

donnchawp committed:
Show the "next preload" message any time it's needed.
Not just when it's close to that preload time.

donnchawp committed:
Remove div tags, not needed

donnchawp committed:
Use wpsc_is_preload_active() for stopping

Add checks to that function for the stop_preload flag, and if the
mutex file is missing then return false.

donnchawp committed:
Simplify stopping the preload. Remove the "preload_cache_stop" flag.

* No need for two "stop" flags to stop the preload.
* Reset the settings immediately, for quicker shutdown.

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/preload_simplification

dilirity commented on projects/plugins/super-cache/wp-cache.php:
```diff
@@ -3624,7 +3632,15 @@ function wpsc_is_preload_active() {
 		return true;
+	} else {
+		return false;
 	}
```
This prevents code after the if statement to get executed.

dilirity commented on projects/plugins/super-cache/wp-cache.php:
```diff
@@ -3605,6 +3632,42 @@ function supercache_admin_bar_render() {
+function wpsc_is_preload_active() {
+	global $cache_path;
+
+	$mutex = $cache_path . 'preload_mutex.tmp';
+	if ( file_exists( $mutex ) ) {
+		return true;
+	} else {
+		return false;
+	}
```
This prevents any code after the if statement from getting executed.

donnchawp replied:
Doh! I'll sort that out.

donnchawp replied:
Fixed by moving the "stop preload" file check above it.

dilirity reviewed (changes requested):
I love how simpler the preload is becoming! Just found one small bug - below.

donnchawp committed:
Catch "stop preload" flag first, then (corrected) mutex check

* This will abort the preload right at the start of the function, if it's cancelled.

donnchawp committed:
Reset settings if preload cancelled, instead of scheduling new task

donnchawp committed:
Fix translators explanation.

donnchawp committed:
No need for $mutex here, only used once

dilirity committed:
Fix missing variable

dilirity reviewed (dismissed):
I think we're good to go :shipit:

donnchawp committed:
Merge branch 'trunk' into update/preload_simplification

donnchawp committed:
Merge with changes in #31017

donnchawp committed:
Use wpsc_is_preload_active() for "Preload now" button check

donnchawp committed:
Set idle status on cancel, and remove "almost cancelled" message.

There's a check on the STOP condition in each loop of the preload, so
it stops very quickly now. No configuration or temp files are updated.

donnchawp committed:
Check the preload counter for the "post" loop as well

donnchawp committed:
Fix typo in option name

dilirity reviewed (changes requested):
Looks good. However, cancelling a preload that hasn't already started still updates the `last preload finished` value.

EDIT:
Let's get this merged - the `last preload finished` can be fixed in another PR.

donnchawp committed:
Just adding whitespace to run checks again

donnchawp committed:
Remove this to trigger checks again

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/preload_simplification
